### PR TITLE
replace pyupgrade+isort+autoflake with ruff

### LINKED ISSUE #101: Replacing black, autoflake and isort with Ruff
With the beta release of Ruff's formatter, ruff now contains the functionality of several of the tools in shed, and runs much faster.

See the ruff docs for the functionality it replaces:

- [isort](https://docs.astral.sh/ruff/rules/#isort-i)
- [pyflakes](https://docs.astral.sh/ruff/rules/#pyflakes-f)
- [black](https://docs.astral.sh/ruff/formatter/#black-compatibility)

I love the pyupgrade and refactor aspects of shed, but it's difficult to use with ruff, as the minor deviations mean I have to run ruff format, run shed, then run ruff format again in order to get the most out of both tools, which then negates a lot of the speed benefits of ruff.

Personally I think ruff should replace those tools within shed, but I would also be happy to have it as an option.

Happy to try for a PR of this if it's something other people are interested in.

### PR DESCRIPTION
**Pull request by jakkdl**
Take 3 after
https://github.com/Zac-HD/shed/compare/master...use-ruff and #102

fixes #101 

Notes:
* The one feature we're losing with replacing autoflake is not exploding star imports. (There's no test for that functionality.)
* when looking for tests I traversed git blame, so bfb380e9538ca1502055cade65e660ffbdba10c8 is now ignored. Unfortunately the commit[s] when @Zac-HD added a ton of carriage returns are in the same commits as adding a bunch of other stuff, so can't ignore them :P
* ruff removing `pass` also removes unnecessary `...`, hence change in `pattern-matching.txt`.

remaining TODOs:
* [x] replace usage of autoflake.is_python_file
  * I can probably just inline it.
* [ ] replace black with `ruff format` (could leave that for a separate PR)
  * figure out a way to not rely on black for autodetecting target version
* figure out minimum ruff version
* open issue about git worktrees
  * I made the test xfail for now
* [x] remove any now-redundant autofixers
  * ~fully done. I'm now working on documenting the remaining ones properly, partly because wtf-why-isn't-that-done-already, and partly so the ruff folks can consider implementing them. But that could be for a different PR.

### CODE REVIEW

**Comment by jakkdl** [comment]
[libcst 1.2 dropped support for 3.8](https://github.com/Instagram/LibCST/releases/tag/v1.2.0). Resolving it by generating `deps.txt` on 3.8, but could also do it by limiting libcst to `<1.2`. Ooooor we can drop 3.8 :wink:

**Comment by jakkdl** [comment]
ugh that actually did break 3.12 somehow. Seems like we might actually have to drop 3.8?

File: CODEMODS.md [hunk 1]
```diff
@@ -0,0 +1,379 @@
...
+set(reversed(iterable))
+sorted(reversed(iterable)) # unsafe to fix, but ruff does
+
+# cases handled by our codemod, but not by ruff
+sorted(sorted(iterable, reverse=True))
+sorted(sorted(iterable, reverse=True))
+sorted(sorted(iterable, reverse=True), reverse=False)
+sorted(sorted(iterable, reverse=False), reverse=True)
+
+# unsafe to fix
```
Comment (Zac-HD) [inline CODEMODS.md:79]:
```suggestion
# unsafe to fix, even if both have the same key function (it might have side-effects)
```

File: CODEMODS.md [hunk 2]
```diff
@@ -0,0 +1,379 @@
...
+
+for _ in range(10):
+    ...
+else:
+    pass
+
+try:
+    1 / 0
+else:
+    pass
```
Comment (Zac-HD) [inline CODEMODS.md:265]: Invalid syntax!

File: src/shed/__init__.py [hunk 3]
```diff
@@ -59,6 +142,8 @@ def shed(
     """Process the source code of a single module."""
     assert isinstance(source_code, str)
     assert isinstance(refactor, bool)
+    # TODO: I guess this required frozenset because of isort compatibility
+    # but we can probably relax that to an iterable[str]
     assert isinstance(first_party_imports, frozenset)
```
Comment (Zac-HD) [inline src/shed/__init__.py:147]: I think it was more because I prefer strict validation unless there's a specific reason not to, and there's the '`str` is an iterable of `str`' footgun.

File: src/shed/__init__.py [hunk 4]
```diff
@@ -146,34 +232,35 @@ def shed(
...
+            capture_output=True,
+        )
+        return sub.stdout
 
-    # Until https://github.com/PyCQA/autoflake/issues/229 is fixed, autoflake
-    # may reorder imports in a way that is incompatible with isort's ordering,
-    # so we may need to re-sort its output.
-    if source_code != pre_autoflake:
-        source_code = run_isort()
+    source_code = run_ruff()
```
Comment (Zac-HD) [inline src/shed/__init__.py:263]: Since we're only running `ruff` once, we can just inline the function definition and save a bit of time 🙂

File: src/shed/_cli.py [hunk 5]
```diff
@@ -45,13 +44,17 @@ def _guess_first_party_modules(cwd: Optional[str] = None) -> FrozenSet[str]:
     return frozenset(
         p
         for p in {Path(base).name} | provides
-        if p.isidentifier() and place_module(p) != "STDLIB"
+        # TODO: isort.place_module is horrendously complicated, but we only use
+        # a fraction of the functionality. So best approach, if we still need
+        # the ability to exclude stdlib modules here, is probably to generate a list of
+        # known stdlib modules - either dynamically or store in a file.
+        if p.isidentifier()  # and place_module(p) != "STDLIB"
```
Comment (Zac-HD) [inline src/shed/_cli.py:51]: Yeah, having a known list seems fine, and we can just take the union of stdlib modules in all non-EOL Python versions.

File: src/shed/_codemods.py [hunk 6]
```diff
@@ -29,7 +29,9 @@ def inner(fn):
         @m.leave(matcher)
         def wrapped(self, original_node, updated_node):
             if not m.matches(updated_node, matcher):
-                return updated_node
+                # given that this doesn't trigger... can we delete it or do we need a new
+                # test case with one of the remaining codemods?
+                return updated_node  # pragma: no cover
```
Comment (Zac-HD) [inline src/shed/_codemods.py:34]: Needs a new test, due to https://github.com/Instagram/LibCST/issues/888.

(or for that to be fixed upstream and then shed requires the new version, but that's a lot of scope creep)

File: tests/recorded/comprehensions/C413.txt [hunk 7]
```diff
@@ -21,16 +22,17 @@ sorted(frozenset('abc'))
...
 
 sorted([2, 3, 1], reverse=False)
-reversed(sorted([2, 3, 1], reverse=False))
-reversed(sorted([2, 3, 1], reverse="abc"))
-reversed(sorted([2, 3, 1], reverse=1 + 2 + c))
-reversed(sorted([2, 3, 1], reverse=1 or 2))
+sorted([2, 3, 1], reverse=True)
+sorted([2, 3, 1], reverse=not "abc")
+sorted([2, 3, 1], reverse=not 1 + 2 + c)
+sorted([2, 3, 1], reverse=not 1 or 2)
```
Comment (Zac-HD) [inline tests/recorded/comprehensions/C413.txt:35]: Hmmm... looks like this can change semantics due to [operator precedence](https://docs.python.org/3/reference/expressions.html#operator-precedence) sometimes.  Mind working out whether that's `shed` or `ruff`, and reporting as appropriate?

File: tests/recorded/issue75.txt [hunk 8]
```diff
@@ -1,3 +1,12 @@
+# The original issue was about shed requiring two passes to format this:
+# on first run autoflake would remove imports, and second run isort would collapse
+# to a single line of imports.
+# Now we have slightly different behaviour: black will add a trailing comma to the
+# import list, which ruff then will respect when removing imports, such that it doesn't
+# collapse the imports to a single line.
```
Comment (Zac-HD) [inline tests/recorded/issue75.txt:6]: suggests that we might want to run `ruff` before `black` then?

**Review by Zac-HD — COMMENTED** [review]
Thanks again @jakkdl, this is looking great!

(especially the documentation so that `ruff` can pick up more of the codemods... extremely valuable and would have sat on my 'todo someday' list ~forever)

File: CODEMODS.md (hunk 2 above)
Comment (jakkdl) [reply CODEMODS.md:265]: haha, oops :sweat_smile:

File: src/shed/_codemods.py (hunk 6 above)
Comment (jakkdl) [reply src/shed/_codemods.py:34]: I spent a while looking through the remaining codemods if any pair of them could trigger this, but didn't manage to find any. But I don't think it's much of a performance overhead so let's just leave it for now in case I missed something, or a new codemod is introduced that could trigger it.
Updated the docstring & comment

File: tests/recorded/comprehensions/C413.txt (hunk 7 above)
Comment (jakkdl) [reply tests/recorded/comprehensions/C413.txt:35]: `shed` didn't dare messing with the reverse key previously. Opened ruff issue: https://github.com/astral-sh/ruff/issues/10335

I can reinstate `shed`s C413 fixer, or we hope for a timely fix.

**Comment by jakkdl** [comment]
> ugh that broke 3.12 somehow. Seems like we might actually have to drop 3.8?

Oh, can just add `libcst<1.2; python_version < "3.9"` to `deps/check.in`.

Mostly unrelated to this PR:
Looking at https://github.com/Zac-HD/shed/actions/runs/8175453287/job/22352606988#step:5:75 says
> `git apply .hypothesis/patches/2024-03-06--c3414d12.patch` to add failing examples to your code.
Not that it matters much here (I think? kinda weird error...), but is it possible to get at that patch in any way when it's run in CI? If not, does hypothesis have a flag for printing that patch in a different way to the log so a user could make use of it?

For funsies I also tried running 3.13 locally, but encountered https://github.com/HypothesisWorks/hypothesis/issues/3916

File: src/shed/__init__.py (hunk 3 above)
Comment (jakkdl) [reply src/shed/__init__.py:147]: ugh, right. I suppose I could change it to `Iterable[str]` and `assert not isinstance(first_party_imports, str)`?

File: src/shed/_cli.py (hunk 5 above)
Comment (jakkdl) [reply src/shed/_cli.py:51]: or well, depends on how thorough we want to be + the level of support for running-version-being-different-from-checked-version

Comment (jakkdl) [reply src/shed/_cli.py:51]: https://pypi.org/project/stdlibs/ sounds pretty great if we want to be maximum thorough

File: tests/recorded/issue75.txt (hunk 8 above)
Comment (jakkdl) [reply tests/recorded/issue75.txt:6]: This messed up pyupgrade ... which made me realize that ruff has pyupgrade rules as well.

Ruff needs to run after codemods (to remove needless pass inserted by `split_assert`), but I don't see any reason why black needs to run before codemods (and running it after fixes one small case in `split_assert`), so changed it to do codemods->ruff->black.

Comment (jakkdl) [reply tests/recorded/issue75.txt:6]:
```py
(
    []
    .append(1)
)
```
when `pointless_parens` tries to remove parens here it doesn't delve into the `Call` node to remove the newline, so removing the parens fail. (and trying to do so sounds like a Bad Idea:tm:) But black unlocks this fix. So a naive codemod->ruff->black will change stuff on second run.

Another case that depends on ordering, which would get improved by not running black first, is
```py
# input
assert (
a # a
and b # b
)
# black -> codemod (status quo)
assert a  # a  # b
assert b
# codemod -> black
assert a  # a
assert b  # b
```

I don't feel like it matters a ton either way, so I suggest keeping status quo of black->codemods->ruff->[black]

**Comment by jakkdl** [comment]
The CI check previously failed because `tox -e check` runs shed, auto-detecting all files in the git repository, which includes `CODEMODS.md`, and `ruff` then wants to modify the "input code" in two of the examples. I managed to get around it by using `'''py` instead of `'''python` (except with backticks) where ruff tries to butt in. Github still seems to display those correctly, but shed doesn't care for them (which it maybe should?)

File: src/shed/_cli.py (hunk 5 above)
Comment (Zac-HD) [reply src/shed/_cli.py:51]: Let's just use that on 3.10+, and vendor the 3.10 list for older versions 👍

File: src/shed/__init__.py (hunk 3 above)
Comment (Zac-HD) [reply src/shed/__init__.py:148]: and then convert it to a frozenset? Sounds good.

File: tests/recorded/comprehensions/C413.txt (hunk 7 above)
Comment (Zac-HD) [reply tests/recorded/comprehensions/C413.txt:35]: Happy to leave this as a `ruff` problem, they do fix things quickly.

**Comment by Zac-HD** [comment]
> ##### Mostly unrelated to this PR:
> Looking at https://github.com/Zac-HD/shed/actions/runs/8175453287/job/22352606988#step:5:75 says
> 
> > `git apply .hypothesis/patches/2024-03-06--c3414d12.patch` to add failing examples to your code.
> 
> Not that it matters much here (I think? kinda weird error...), but is it possible to get at that patch in any way when it's run in CI? If not, does hypothesis have a flag for printing that patch in a different way to the log so a user could make use of it?

If you've [configured artifact upload for that directory, yes](https://docs.github.com/en/actions/using-workflows/storing-workflow-data-as-artifacts).  It's useful to print them though, because if there's a printable example we already show them - in this case [here](https://github.com/Zac-HD/shed/actions/runs/8175453287/job/22352606988#step:5:55) with `source_code='#\rµ',`.

**Review by Zac-HD — APPROVED** [review]
I think this is a nice improvement over the status quo, so merging now.

If you want to make a follow-up PR with the stdlib list and optionally the iterable/frozenset handling, plus a new version number, I'll release shortly 🎉

File: tests/recorded/comprehensions/C413.txt (hunk 7 above)
Comment (jakkdl) [reply tests/recorded/comprehensions/C413.txt:35]: And indeed it was!

File: src/shed/__init__.py (hunk 3 above)
Comment (jakkdl) [reply src/shed/__init__.py:148]: It was previously passed to `isort` as a frozenset, but now we only use it as a parameter to ruff - where I currently do (the equivalent of) `str(list(first_party_imports))` for formatting it. So if we'd only accept a `list[str]` we don't need to convert the type... not that it really matters though.
